crates/router-controller/src/request_tracker.rs

```diff
@@ -0,0 +1,93 @@
+use std::time::{Duration, Instant};
+use std::collections::HashMap;
+
+use crate::controller::{Url, Dataset};
+
+
+#[derive(Clone)]
+struct RequestInfo {
+    url: Url,
+    expiration: Instant,
+}
+
+
+impl RequestInfo {
+    fn new(val: Url, duration: Duration) -> RequestInfo {
+        RequestInfo {
+            url: val,
+            expiration: Instant::now() + duration,
+        }
+    }
+
+    fn is_expired(&self) -> bool {
+        Instant::now() > self.expiration
```

Comment:
As far as I know, that's a syscall. Syscalls are never cheap inside critical loops.